File: clients/java/cldf-java/src/main/java/io/cldf/models/enums/FinishType.java
package io.cldf.models.enums;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

/** How the climb was completed. */
public enum FinishType {
  // Common finish types for both boulder and route
  FLASH("flash"),
  TOP("top"),
  REPEAT("repeat"),
  PROJECT("project"),
  ATTEMPT("attempt"), // Attempting the climb but falling or giving up

  // Route-specific finish types (onsight and redpoint are route-only)
  ONSIGHT("onsight"),
  REDPOINT("redpoint");

  // Note: flash, repeat, project, and attempt appear in both boulder and route schemas
  // but we define them once here. The validation should be done at a higher level
  // based on the climb type.

  private final String value;

  FinishType(String value) {
    this.value = value;
  }

  @JsonValue
  public String getValue() {
    return value;
  }

  @JsonCreator
  public static FinishType fromValue(String value) {
    if (value == null) {
      return null;
    }
    for (FinishType type : values()) {
      if (type.value.equalsIgnoreCase(value)) {
        return type;
      }
    }
    throw new IllegalArgumentException("Unknown finish type: " + value);
  }
}

File: clients/dart/cldf/lib/models/enums/finish_type.dart
import 'package:json_annotation/json_annotation.dart';
import 'climb_type.dart';

@JsonEnum(valueField: 'value')
enum FinishType {
  // Boulder finish types
  flash('flash'),
  top('top'),

  // Route finish types
  onsight('onsight'),
  redpoint('redpoint'),

  // Common finish types
  repeat('repeat'),
  project('project'),
  attempt('attempt');

  final String value;
  const FinishType(this.value);

  /// Check if this is a boulder-specific finish type
  bool get isBoulderFinish => this == flash || this == top;

  /// Check if this is a route-specific finish type
  bool get isRouteFinish => this == onsight || this == redpoint;

  /// Check if this finish type is valid for the given climb type
  bool isValidForClimbType(ClimbType climbType) {
    if (climbType == ClimbType.boulder) {
      return isBoulderFinish || [repeat, project, attempt].contains(this);
    } else {
      return isRouteFinish || [flash, repeat, project, attempt].contains(this);
    }
  }
}
